### DISCUSSION TITLE
Undefined 'dvid' integer values in data: 1, 2, 3

### DISCUSSION THREAD

**Discussion by Dstix** [nlmixr2/rxode2#373] [Q&A] [answered]
I have an oral 2-comp indirect response model with the concentration compartment labeled as `cmt = cp`, the effect compartment labeled as `cmt = eff`, and the gut compartment labeled as `cmt = gut`. In my data set, the `dvid` I have assigned to them are 1, 2, and 3 respectively. When I run my fit, the `cmt` is output, but not the `dvid`. When I run the command:

`clean_fit <- augPred(fit)`

I get the following error:

`'DVID'/'CMT' translation:
  dvid modeled cmt
1    1           5
2    2           4
'CMT': 17
Error in rxSolveSEXP(object, .ctl, .nms, .xtra, params, events, inits,  : 
  'dvid'->'cmt' or 'cmt' on observation record on a undefined compartment (use 'cmt()' 'dvid()') id: 1 row: 0
In addition: Warning message:
Undefined 'dvid' integer values in data: 1, 2, 3`

**Comment by mattfidler** [comment]
Well, `dvid` is converted to `cmt` so it not being present is not terribly suprising, though I'm unsure how to reproduce and help with the data.  

Could you add a reproducible example so I can help you with this issue?

**Reply by Dstix** [reply]
This might be tricky as I am using internal data...

**Comment by billdenney** [comment]
@Dstix , can you share the model function (changing proprietary numbers and names) along with the unique values in the dvid column?  That may give us some hints.

**Reply by Dstix** [reply]
Sure, I can do that...

Unique dvid values:

`unique(pk_pd_nl$dvid)
[1] 3 2 1`

indresp <- function() {
  ini({
    tv_cl <- log(7); label("log Cl")
    tv_q <- log(3); label("log Q")
    tv_v2 <- log(60); label("log V2")
    tv_v3 <- log(250); label("log V3")
    tv_ka <- log(0.1); label("log ka")
    tv_emax <- 0.6; label("eMax")
    tv_ic50 <- log(3); label("log ic50")
    tv_kdeg <- log(6); label("log kdeg")
    tv_e0 <- log(350); label("log baseline")
    eta_cl ~ 0.2
    eta_ka ~ 0.2
    eta_e0 ~ 0.1
    prop_err <- 0.2
    add_err <- 300
  })
  model({
    cl <- exp(tv_cl + eta_cl)
    v2 <- exp(tv_v2)
    v3 <- exp(tv_v3)
    q  <- exp(tv_q)
    ka <- exp(tv_ka + eta_ka)    
    emax <- tv_emax
    ic50 <- exp(tv_ic50)    
    kdeg <- exp(tv_kdeg)    
    e0 <- exp(tv_e0 + eta_e0)
    eff(0) <- e0
    ksyn <- kdeg * e0    
    cp = central * 1000 / v2    
    d/dt(gut) = -ka * gut    
    d/dt(central) = ka * gut - cl / v2 * central + q/v3 * periph - q/v2 * central
    d/dt(periph) = q/v2 * central - q/v3 * periph
    d/dt(eff) = ksyn * (1 - cp * emax / (cp + ic50)) - kdeg * eff    
    cp ~ prop(prop_err)
    eff ~ add(add_err)
  })
}

**Comment by billdenney** [comment, marked as answer]
And, what are the unique dvid and cmt values in your data?  (`unique(data[, c("cmt", "dvid")])`)

**Reply by Dstix** [reply]
Here you are:

```
pk_pd_nl %>% distinct(cmt,  dvid)

# A tibble: 3 × 2
  cmt    dvid
  <chr> <dbl>
1 gut       3
2 eff       2
3 cp        1
```
Now that you mention it, when I solve the model, it seems to make a new compartment for cp (cmt = 5) and for eff (cmt = 4). When I type:

`clean_fit <- augPred(fit)`

I get:
```
'DVID'/'CMT' translation:
  dvid modeled cmt
1    1           5
2    2           4
'CMT': 17
Error in rxSolveSEXP(object, .ctl, .nms, .xtra, params, events, inits,  : 
  'dvid'->'cmt' or 'cmt' on observation record on a undefined compartment (use 'cmt()' 'dvid()') id: 1 row: 0
In addition: Warning message:
Undefined 'dvid' integer values in data: 1, 2, 3
```
I know I am missing something simple...

**Reply by billdenney** [reply]
@Dstix, unless I'm missing something, you don't need the `dvid` column in your data.  I think that your model will work without it.  (https://nlmixr2.org/articles/multiple-endpoints.html#dvid-vs-cmt-which-one-is-used)

Also, I think that your model would also be fixed if you used text instead of numeric values for dvid (if you prefer to keep the dvid).  FYI, to simplify these types of issues, I only use text values for `cmt` in my models (and I very rarely use `dvid`).

**Reply by Dstix** [reply]
I completely removed the dvid column from my input data set `pk_pd_nl %<>% select(-dvid)` (I was just using to see if some of these problems resolved themselves). Here is the error I get (nearly the same as before):

```
'DVID'/'CMT' translation:
  dvid modeled cmt
1    1           5
2    2           4
'CMT': 17
Error in rxSolveSEXP(object, .ctl, .nms, .xtra, params, events, inits,  : 
  'dvid'->'cmt' or 'cmt' on observation record on a undefined compartment (use 'cmt()' 'dvid()') id: 1 row: 0
```

**Reply by billdenney** [reply]
@Dstix, Ah, I missed an important detail.  (Sorry, I don't often use multi-endpoint models in nlmixr2 yet.)  Try the following:

* In your data: set your "cmt" column where it is currently "cp" to "central".  (`ifelse(cmt %in% "cp", "central", cmt)`)
* In your model, specify that cp is in the central compartment by changing your estimation line: `cp ~ prop(prop_err) | central`

**Reply by Dstix** [reply]
That did the trick. When I ran it before, nlmixr2 needed dvid or it would throw an error. When I removed the dvid column, corrected the "cp" cmt to the "central" compartment and did `cp ~ prop(prop_err) | central` augPred worked great. Thanks for your patience!

**Comment by mattfidler** [comment]
To be clear do you mean you have a successful minimization following by an unsuccessful `augPred()`?

**Reply by Dstix** [reply]
I did but I specified the central compartment as "cp" instead of "central". Using Bill's corrections, it works now.

**Reply by mattfidler** [reply]
Great!  Thanks Bill
